File: server/common/message_creator.py
import logging
from common.utils import Bet


def create_encoded_message(response_code, client_ID):
    encoded_data = ""
    formatted_data = "| {} | {} | {} |".format(response_code, client_ID, encoded_data)
    formatted_data_bytes = formatted_data.encode('utf-8')
    buf = bytearray()
    buf.extend(formatted_data_bytes)
    try:
        encoded_data_big_endian = bytes(buf)

        logging.info("action: encode | result: success | client_id: %s | data: %s",
                     client_ID,
                     encoded_data)

        return encoded_data_big_endian
    except Exception as e:
        logging.error(
            "action: encode | result: fail | client_id: %s | error: %s | data: %s",
            client_ID,
            str(e),
            encoded_data,
        )
        return None


def decode_message(received_data):
    try:
        parts = received_data.decode('utf-8').split('|')
        if len(parts) != 5:
            raise ValueError("Formato inválido")

        message_code = int(parts[1].strip())
        client_ID = parts[2].strip()
        bet_data = parts[3].split(',')
        bet = Bet(client_ID, bet_data[0].strip(), bet_data[1].strip(), bet_data[2].strip(), bet_data[3].strip(), bet_data[4].strip())
        logging.info("action: decode | result: success | client_id: %s", client_ID)

        return message_code, client_ID, bet
    
    except Exception as e:
        logging.error(
            "action: decode | result: fail | client_id: %s | error: %s",
            client_ID,
            str(e),
        )
        return None, None, None

File: client/common/message-creator.go
package common

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"strconv"
	"strings"

	log "github.com/sirupsen/logrus"
)

// El mensaje tiene la forma: ` | código | agencia | datos | `

func CreateEncodedMessage(client *Client, messageCode int, batch [][]string) []byte {
	encodedData := ""
	formattedData := ""

	for i := 0; i < len(batch); i++ {
		encodedData = fmt.Sprintf(
			"%s,%s,%s,%s,%s",
			batch[i][0],
			batch[i][1],
			batch[i][2],
			batch[i][3],
			batch[i][4],
		)
		formattedData += fmt.Sprintf("| %s | %s | %s |", strconv.Itoa(messageCode), client.config.ID, encodedData)
	}

	if messageCode == 3 {
		formattedData += fmt.Sprintf("| %s | %s | %s |", strconv.Itoa(messageCode), client.config.ID, encodedData)
	}

	formattedDataBytes := []byte(formattedData)

	buf := new(bytes.Buffer)
	err := binary.Write(buf, binary.BigEndian, formattedDataBytes)

	if err != nil {
		log.Fatalf(
			"action: encode | result: fail | client_id: %v | error: %v | data: %v",
			client.config.ID,
			err,
			encodedData,
		)
	}

	formattedDataBigEndian := buf.Bytes()

	log.Infof("action: encode | result: success | client_id: %v | data: %v",
		client.config.ID,
		encodedData,
	)

	return formattedDataBigEndian
}

func DecodeMessage(client *Client, receivedData []byte) int {
	formattedData := string(receivedData)

	log.Infof("action: decode | result: success | client_id: %v | data: %v",
		client.config.ID,
		formattedData,
	)

	parts := strings.Split(formattedData, "|")
	if len(parts) != 5 {
		log.Fatalf(
			"action: decode | result: fail | client_id: %v | error: formato invalido",
			client.config.ID,
		)
	}

	messageCode, _ := strconv.Atoi(strings.TrimSpace(parts[1]))

	log.Infof("action: decode | result: success | client_id: %v | messageCode: %d",
		client.config.ID,
		messageCode,
	)

	return messageCode
}
